### PR TITLE
Preparation for MultiAccount support

### PR DESCRIPTION
**Pull request by bblommers**
This PR is an first step towards support for multi-accounts, by changing the underlying data structure to enable this feature.

### Current implementation:
Most services use an instance of the `BackendDict`-class to persist the state.
The BackendDict overrides a dictionary, and stores data in the following format:
```
{
  "us-east-1": ServiceBackend("us-east-1"),
  "us-east-2": ServiceBackend(..), 
  ..
}
```

Because global services such as S3 do not have any regions, those backends were persisted as straightforward dictionaries:
```
{"global": S3Backend()}
```

### Proposed implementation:
All services now use the `BackendDict`, with an option to specify whether this is a global service.
The underlying data structure of the BackendDict now looks like this:
```
{
  "account_id": {
    "us-east-1": ServiceBackend(region_name="..", account_id=".."),
    ..,
  },
  ..
}
```

**Benefits of this approach:**
 - There is a logical separation between accounts
 - The BaseBackend becomes responsible for setting/resetting the `region_name` and `account_id`, reducing code duplication
 - Because of that, each ServiceBackend has direct access to it's account id, and a user longer has to worry about overriding it on a case-by-case basis

### Noticeable aspects of this PR:
There is no change in functionality.
 - The BackendDict will always return the same account_id for now, regardless of what the user specifies.
 - No ServiceBackend uses `self.account_id`, but continues to import `get_account_id`

A lot of users access the internal API using `backend_dict["us-east-1"].model.attribute = ".."`.
The `__getitem__`-method of BackendDict has been overridden to still support this, enabling backwards compatibility

The meat of this PR lies in `moto/core/utils/BackendDict` - all other changes are a logical consequence of the changes to that class.

### Future PR's:
Because of the massive scope of this change, the current PR only prepares multi-account support, but without any functional changes.
A future PR is necessary to then unlock this functionality. This PR would consist of the following changes:
 - Ensure that each ServiceResponse has access to the AccountID for that specific request, using either `get_account_id` or IAM
 - Ensure that each ServiceResponse retrieves the correct ServiceBackend, using `backend_dict[account_id][region_name]`
 - Change all services to use `self.account_id`, instead of `get_account_id()`
 - Remove the option to access `backend_dict[region_name]` directly, as we would  have no context about which account we want to use

Because this PR would be a breaking change, it would have to be released as part of the next major release.

Related PR's:
#4699 introduced the BackendDict in the first place
#5098 added another way to override AccountID

### TODO:
Need to verify that the scaffold-script still creates the expected templates.

### CODE REVIEW

**Comment by bblommers** [comment]
@whummer @thrau FYI. As far as I can tell, this should not have an impact on on LocalStack. Let me know if you need some time to review/test this though.

**Comment by thrau** [comment]
this looks great @bblommers! excited to see that multi-account is coming to moto :-) i'm sure this will simplify things also on our end for multi-account. i haven't fully understood the ramifications of the change yet, but as you said it looks like this won't have much impact since it's backwards compatible. it may affect some monkeypatches to backends but we can iron those out.
we can try to run a localstack ci build with the changes and see what happens.

**Review by whummer — COMMENTED** [review]
Wow, that is a massive set of changes, big kudos @bblommers and thanks for keeping us in the loop, 🙏 really appreciated! Great to see that you were also able to simplify/streamline some of the `__init__` super calls, as well as default `reset(..)` implementations. 👍 

My only question would be around hardcoding account ID `"123456789012"` in the new logic in `BackendDict` (see comments) - other than that, this is a huge enhancement and I think we can mostly work around the changes in here, as @thrau mentioned. 👍 Looking forward to seeing this in action! 🥳

File: moto/core/utils.py [hunk 1]
```diff
@@ -409,27 +409,120 @@ def extract_region_from_aws_authorization(string):
...
+        """
+        Possible data access patterns:
+          backend_dict[account_id][region_name]
+          backend_dict[region_name]
+          backend_dict[unknown_region]
+
+        The latter two will be phased out in the future, and we can remove this method.
+        """
+        if re.match(r"[0-9]+", account_id_or_region):
+            self._create_account_specific_backend("123456789012")
```
[the remark this answers is not in the document]
Comment (bblommers) [reply moto/core/utils.py:445]: This was done on purpose - hardcoding the ID keeps everything in a single account for now, ensuring that the current behaviour doesn't change.

The follow-up PR will definitely change this to `get_account_id`, but that will need a few other changes - and many more tests.
Wrapping everything in one PR would make it even more unwieldy and impossible to review, that's why I chose to break it up in two steps: 1 PR to prepare, and 1 PR to actually change the behaviour.

[the remark this answers is not in the document]
Comment (whummer) [reply moto/core/utils.py:445]: Understood. That makes a lot of sense, thanks for clarifying! 👍

**Pull request merged**